src/runtime-rs/crates/resource/src/network/dan.rs

```diff
@@ -0,0 +1,371 @@
+                NetworkInfoFromDan::new(device)
+                    .await
+                    .context("Network info from DAN")?,
+            );
+
+            entity_list.push(NetworkEntity {
+                endpoint,
+                network_info,
+            })
+        }
+
+        Ok(Self { entity_list })
+    }
+}
+
+#[async_trait]
+impl Network for Dan {
+    async fn setup(&self) -> Result<()> {
+        let inner = self.inner.read().await;
+        for e in inner.entity_list.iter() {
+            e.endpoint.attach().await.context("Attach")?;
+        }
+        Ok(())
+    }
+
+    async fn interfaces(&self) -> Result<Vec<agent::Interface>> {
+        let inner = self.inner.read().await;
+        let mut interfaces = vec![];
+        for e in inner.entity_list.iter() {
+            interfaces.push(e.network_info.interface().await.context("Interface")?);
+        }
+        Ok(interfaces)
+    }
+
+    async fn routes(&self) -> Result<Vec<agent::Route>> {
+        let inner = self.inner.read().await;
+        let mut routes = vec![];
+        for e in inner.entity_list.iter() {
+            let mut list = e.network_info.routes().await.context("Routes")?;
+            routes.append(&mut list);
+        }
+        Ok(routes)
+    }
+
+    async fn neighs(&self) -> Result<Vec<agent::ARPNeighbor>> {
+        let inner = self.inner.read().await;
+        let mut neighs = vec![];
+        for e in &inner.entity_list {
+            let mut list = e.network_info.neighs().await.context("Neighs")?;
+            neighs.append(&mut list);
+        }
+        Ok(neighs)
+    }
+
+    async fn save(&self) -> Option<Vec<EndpointState>> {
+        let inner = self.inner.read().await;
+        let mut ep_states = vec![];
+        for e in &inner.entity_list {
+            if let Some(state) = e.endpoint.save().await {
+                ep_states.push(state);
+            }
+        }
+        Some(ep_states)
+    }
+
+    async fn remove(&self, h: &dyn Hypervisor) -> Result<()> {
+        let inner = self.inner.read().await;
+        for e in inner.entity_list.iter() {
+            e.endpoint.detach(h).await.context("Detach")?;
+        }
+        Ok(())
+    }
+}
+
+/// Directly attachable network config
+#[derive(Debug)]
+pub struct DanConfig {
+    pub dan_conf_path: PathBuf,
+}
+
+/// Directly attachable network device
+/// This struct is serilized from a file containing devices information,
+/// sent from CNI plugins.
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub(crate) struct DanDevice {
+    // Name of device (interface name on the guest)
+    pub(crate) name: String,
+    // Mac address of interface on the guest, if it is not specified, a
+    // private address is generated as default.
+    #[serde(default = "generate_private_mac_addr")]
+    pub(crate) guest_mac: String,
+    // Device
+    pub(crate) device: Device,
+    // Network info
+    pub(crate) network_info: NetworkInfo,
+}
```

Comment:
Mybe we should also support cni pass a netns path, since the tap device would be assigned in a specific netns instead of in the init netns.

Reply:
I've changed the struct of DAN config. The netns is supported to pass by the JSON file now.